### PR TITLE
Address printing inconsistency between float and complex tensors

### PR DESCRIPTION
**Pull request by choidongyeon**
See issue [#33494 Complex number printing inconsistent with float](https://github.com/pytorch/pytorch/issues/33494).

Changes introduces an optional argument in Formatter's ```format``` function to discern whether a tensor is a float tensor or not. This way, there is consistency between float tensors and complex tensors so that the complex tensors print in the same manner as float tensors:

- Only a decimal point and no zeros for integer values.
- Trailing zeros only if the value is truly a float.
- White space introduced to fill the gap so that +/- symbols and commas align.

Here are some example outputs.

```
print(torch.zeros((2,2), dtype=torch.float64))
```
yields
```
tensor([[0., 0.],
        [0., 0.]], dtype=torch.float64)
```

```
print(torch.zeros((2,2), dtype=torch.complex64))
```
previously yielded
```
tensor([[(0.0000 + 0.0000j), (0.0000 + 0.0000j)],
        [(0.0000 + 0.0000j), (0.0000 + 0.0000j)]], dtype=torch.complex64)
```
and now yields
```
tensor([[(0 + 0.j), (0 + 0.j)],
        [(0 + 0.j), (0 + 0.j)]], dtype=torch.complex64)
```
This new print version is more consistent with float tensor's pretty print.

The following example mixes integer and decimals:
```
print(torch.tensor([[1 + 1.340j, 3 + 4j], [1.2 + 1.340j, 6.5 + 7j]], dtype=torch.complex64))
```
This yields:
```
tensor([[                     (1.0000 + 1.3400j),
                              (3.0000 + 4.0000j)],
        [                     (1.2000 + 1.3400j),
                              (6.5000 + 7.0000j)]], dtype=torch.complex64)
```

The following example 
```
torch.tensor([1,2,3,4.5])
```
yields
```
tensor([1.0000, 2.0000, 3.0000, 4.5000]) .
```

### CODE REVIEW

**Comment by choidongyeon** [comment]
@anjali411, does this address the printing inconsistency to your liking?

File: torch/_tensor_str.py [hunk 1]
```diff
@@ -145,7 +145,13 @@ def format(self, value):
                 ret = ('{{:.{}f}}').format(PRINT_OPTS.precision).format(value)
         elif self.complex_dtype:
             p = PRINT_OPTS.precision
-            ret = '({{:.{}f}} {{}} {{:.{}f}}j)'.format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
+            # format real and imaginary values according to type
+            real_val = '({{:.0f}}.    '.format(p).format(value.real) if value.real.is_integer() \
```
Comment (anjali411) [inline torch/_tensor_str.py:149]: ```real_val = '({{:.0f}}.'.format(p).format(value.real) if value.real.is_integer() \```

Comment (anjali411) [reply torch/_tensor_str.py:149]: I think we should remove the extra white space

File: torch/_tensor_str.py [hunk 2]
```diff
@@ -145,7 +145,13 @@ def format(self, value):
...
+                else '({{:.{}f}}'.format(p).format(value.real)
+            imag_val = '{{:.0f}}.j    )'.format(p).format(value.imag) if value.imag.is_integer() \
```
Comment (anjali411) [inline torch/_tensor_str.py:151]: ```imag_val = '{{:.0f}}.j)'.format(p).format(value.imag) if value.imag.is_integer() \```

same as above

**Comment by anjali411** [comment]
I think this behavior is in line with what we do for floating point tensors:

```
print(torch.tensor([[1 + 1.340j, 3 + 4j], [1.2 + 1.340j, 6.5 + 7j]], dtype=torch.complex64))

tensor([[                     (1.0000 + 1.3400j),
                              (3.0000 + 4.0000j)],
        [                     (1.2000 + 1.3400j),
                              (6.5000 + 7.0000j)]], dtype=torch.complex64)
```

for example, this: ``` >>> torch.tensor([1,2,3,4.5]) ``` 

gives

```tensor([1.0000, 2.0000, 3.0000, 4.5000])```

File: torch/_tensor_str.py (hunk 1 above)
Comment (choidongyeon) [reply torch/_tensor_str.py]: Addressed in PR update.

File: torch/_tensor_str.py (hunk 2 above)
Comment (choidongyeon) [reply torch/_tensor_str.py]: Addressed in PR update.

**Comment by choidongyeon** [comment]
Updated the PR description with current outputs.

**Comment by choidongyeon** [comment]
Based on the CircleCI build failures summary and remediations, it seems that of the six failing tests, five of them are due to upstream breakages. My understanding based on other PRs I've read is that the sixth failure (rocmdeb) is flaky. Is there anything else I should address?

File: torch/_tensor_str.py [hunk 3]
```diff
@@ -145,7 +145,11 @@ def format(self, value):
                 ret = ('{{:.{}f}}').format(PRINT_OPTS.precision).format(value)
         elif self.complex_dtype:
             p = PRINT_OPTS.precision
-            ret = '({{:.{}f}} {{}} {{:.{}f}}j)'.format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
+            # format real and imaginary values according to type
+            if not is_float_tensor:
```
Comment (anjali411) [inline torch/_tensor_str.py:149]: what are you trying to do here?

Strictly speaking, is_floating_point() returns false for complex, so it's misleading to have is_float_tensor check and do something for it under complex branch

Comment (choidongyeon) [reply torch/_tensor_str.py:149]: You said we only want to truncate the zeros only in cases where all elements of the complex are integers. This checks to see if the complex tensor has any float components. I can change the variable name, but that's what it does.

Comment (anjali411) [reply torch/_tensor_str.py]: hmm maybe change it to has_non_zero_decimal_val ?

Comment (choidongyeon) [reply torch/_tensor_str.py]: Oh, I changed it to ```complex_contains_float```.  Can change it to```has_non_zero_decimal_val ``` if you prefer that though, let me know!

Comment (anjali411) [reply torch/_tensor_str.py]: yeah let's change it to ```has_non_zero_decimal_val``` :)

Comment (choidongyeon) [reply torch/_tensor_str.py]: Okay! Doing it right now.

Comment (choidongyeon) [reply torch/_tensor_str.py]: Updated PR, with values consistent as shown in the PR description.

**Review by anjali411 — APPROVED** [review]

**Comment by anjali411** [comment]
Hey @choidongyeon thanks for the PR! :D

I think that you don't have the latest commits checked out for some submodules and your branch is also behind master by a bunch of commits -- which explains the failing tests.
can you run:
```
git pull https://github.com/pytorch/pytorch.git
git submodule sync; git submodule deinit -f . && git submodule update --init --recursive
git push origin master
``` and then rebase your branch and update the PR>

**Comment by anjali411** [comment]
also did you look into why we have the extra space in printing, for eg.
```
tensor([[                     (1.0000 + 1.3400j),
                              (3.0000 + 4.0000j)],
        [                     (1.2000 + 1.3400j),
                              (6.5000 + 7.0000j)]], dtype=torch.complex64)
```

is it because we are not using masked_select?

**Comment by choidongyeon** [comment]
@anjali411 Thanks for merging the changes from master in for me. Regarding your question about why we have the extra spaces - I haven't investigated too deeply but I know that it has to do with ```self.max_width```. At t he moment, format returns: ```return (self.max_width - len(ret)) * ' ' + ret``` where ```self.max_width``` seems to be set in the Formatter's constructor. One way to address the issue would be to reverse the order of the formatting (assume integer, then filter out cases with actual decimals):

```
        elif self.complex_dtype:
            p = PRINT_OPTS.precision
            ret = "({{:.0f}} {{}} {{:.0f}}.j)".format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
            if has_non_zero_decimal_val:
                # complex tensor contains decimal values elements only
                return '({{:.{}f}} {{}} {{:.{}f}}j)'.format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
        else:
            ret = '{}'.format(value)
        return (self.max_width - len(ret)) * ' ' + ret
```

Then, our return values will look like this:

```
tensor([[0., 0.],
        [0., 0.]], dtype=torch.float64) 

tensor([[(0 + 0.j), (0 + 0.j)],
        [(0 + 0.j), (0 + 0.j)]], dtype=torch.complex64) 

tensor([1.2000, 1.3400], dtype=torch.float64) 

tensor([(1.2000 + 1.3400j)], dtype=torch.complex64) 

tensor([[(1.0000 + 1.3400j),
         (3.0000 + 4.0000j)],
        [(1.2000 + 1.3400j),
         (6.5000 + 7.0000j)]], dtype=torch.complex64) 

tensor([1.0000, 2.0000, 3.0000, 4.5000]) 
```
The only thing is, I feel like there should be a more elegant solution to preserve the current form of the code (one return statement). I can create another issue and look into it separately. Or just push the proposed change here in. Let me know!

**Comment by anjali411** [comment]
> @anjali411 Thanks for merging the changes from master in for me. Regarding your question about why we have the extra spaces - I haven't investigated too deeply but I know that it has to do with `self.max_width`. At t he moment, format returns: `return (self.max_width - len(ret)) * ' ' + ret` where `self.max_width` seems to be set in the Formatter's constructor. One way to address the issue would be to reverse the order of the formatting (assume integer, then filter out cases with actual decimals):
> 
> ```
>         elif self.complex_dtype:
>             p = PRINT_OPTS.precision
>             ret = "({{:.0f}} {{}} {{:.0f}}.j)".format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
>             if has_non_zero_decimal_val:
>                 # complex tensor contains decimal values elements only
>                 return '({{:.{}f}} {{}} {{:.{}f}}j)'.format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
>         else:
>             ret = '{}'.format(value)
>         return (self.max_width - len(ret)) * ' ' + ret
> ```
> 
> Then, our return values will look like this:
> 
> ```
> tensor([[0., 0.],
>         [0., 0.]], dtype=torch.float64) 
> 
> tensor([[(0 + 0.j), (0 + 0.j)],
>         [(0 + 0.j), (0 + 0.j)]], dtype=torch.complex64) 
> 
> tensor([1.2000, 1.3400], dtype=torch.float64) 
> 
> tensor([(1.2000 + 1.3400j)], dtype=torch.complex64) 
> 
> tensor([[(1.0000 + 1.3400j),
>          (3.0000 + 4.0000j)],
>         [(1.2000 + 1.3400j),
>          (6.5000 + 7.0000j)]], dtype=torch.complex64) 
> 
> tensor([1.0000, 2.0000, 3.0000, 4.5000]) 
> ```
> 
> The only thing is, I feel like there should be a more elegant solution to preserve the current form of the code (one return statement). I can create another issue and look into it separately. Or just push the proposed change here in. Personally, I think we should close this issue out by merging it, but let me know!

Hi @choidongyeon I see. I agree we should figure out a more elegant solution for this. I'll merge this PR. would you like to create a follow up issue and work on it?

**Comment by choidongyeon** [comment]
> > @anjali411 Thanks for merging the changes from master in for me. Regarding your question about why we have the extra spaces - I haven't investigated too deeply but I know that it has to do with `self.max_width`. At t he moment, format returns: `return (self.max_width - len(ret)) * ' ' + ret` where `self.max_width` seems to be set in the Formatter's constructor. One way to address the issue would be to reverse the order of the formatting (assume integer, then filter out cases with actual decimals):
> > ```
> >         elif self.complex_dtype:
> >             p = PRINT_OPTS.precision
> >             ret = "({{:.0f}} {{}} {{:.0f}}.j)".format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
> >             if has_non_zero_decimal_val:
> >                 # complex tensor contains decimal values elements only
> >                 return '({{:.{}f}} {{}} {{:.{}f}}j)'.format(p, p).format(value.real, '+-'[value.imag < 0], abs(value.imag))
> >         else:
> >             ret = '{}'.format(value)
> >         return (self.max_width - len(ret)) * ' ' + ret
> > ```
> > 
> > 
> > Then, our return values will look like this:
> > ```
> > tensor([[0., 0.],
> >         [0., 0.]], dtype=torch.float64) 
> > 
> > tensor([[(0 + 0.j), (0 + 0.j)],
> >         [(0 + 0.j), (0 + 0.j)]], dtype=torch.complex64) 
> > 
> > tensor([1.2000, 1.3400], dtype=torch.float64) 
> > 
> > tensor([(1.2000 + 1.3400j)], dtype=torch.complex64) 
> > 
> > tensor([[(1.0000 + 1.3400j),
> >          (3.0000 + 4.0000j)],
> >         [(1.2000 + 1.3400j),
> >          (6.5000 + 7.0000j)]], dtype=torch.complex64) 
> > 
> > tensor([1.0000, 2.0000, 3.0000, 4.5000]) 
> > ```
> > 
> > 
> > The only thing is, I feel like there should be a more elegant solution to preserve the current form of the code (one return statement). I can create another issue and look into it separately. Or just push the proposed change here in. Personally, I think we should close this issue out by merging it, but let me know!
> 
> Hi @choidongyeon I see. I agree we should figure out a more elegant solution for this. I'll merge this PR. would you like to create a follow up issue and work on it?

Sounds like a plan! Will do that right now.

**Pull request closed without merging**